cpu/esp_common/esp-riscv/thread_arch.c

```diff
@@ -0,0 +1,216 @@
+ * in `$(RIOTCPU)/risc_common/thread_arch.c`, which is under the following
+ * copyright:
+ *
+ * Copyright (C) 2017, 2019 Ken Rabold, JP Bonn
+ *
+ * Modifications:
+ * - For compatibility with ESP-IDF, the context frame is defined as
+ *   used by the ESP-IDF.
+ * - The `STACK_MARKER` is not used to identify the top of the stack.
+ *   Instead, the stack is initialized in the `thread_create` function in
+ *   `$(RIOTBASE)/core/thread.c` as expected in the `thread_measure_stack_free`
+ *   function.
+ * - Support for `__global_pointer$` and TLS has been added.
+ */
+char* thread_stack_init(thread_task_func_t task_func, void *arg, void *stack_start, int stack_size)
+{
+    _Static_assert(sizeof(context_switch_frame_t) == 32 * sizeof(uint32_t),
+                   "context frame has to store 32 registers");
+
+    context_switch_frame_t *sf;
+    uint8_t *stk_top;
+
+    /* calculate the top of the stack */
+    stk_top = (uint8_t *)stack_start + stack_size;
+    /* per ABI align stack pointer to 16 byte boundary. */
+    stk_top = (uint8_t *)((uintptr_t)stk_top & ~((uintptr_t)0xf));
+
+    /* prepare thread local storage and the thread pointer */
+    extern uint32_t __global_pointer$;
+
+#if !defined(RISCV_NO_RELAX)
+    extern char _thread_local_start, _thread_local_end, _flash_rodata_start;
+
+    uint8_t *_local_start;
+    uint32_t _local_size = (uint32_t)(&_thread_local_end - &_thread_local_start);
+
+    _local_size = ALIGNUP(0x10, _local_size);
+    stk_top -= _local_size;
+
+    _local_start = stk_top;
+    memcpy(_local_start, &_thread_local_start, _local_size);
+
+    uint8_t *tp;
+    tp = _local_start - (&_thread_local_start - &_flash_rodata_start);
+#else
+    uint8_t *tp = NULL;
+#endif
+
+    /* reserve space for the stack frame. */
+    stk_top = stk_top - sizeof(*sf);
+
+    /* populate the stack frame with default values for starting the thread. */
+    sf = (struct context_switch_frame *)((uintptr_t)stk_top);
+
+    /* Clear stack frame */
+    memset(sf, 0, sizeof(*sf));
+
+    /* set initial reg values */
+    sf->mepc = (uint32_t)task_func;
+    sf->a0 = (uint32_t)arg;
+    sf->gp = (uint32_t)__global_pointer$;
+    sf->tp = (uint32_t)tp;
+    /* if the thread exits go to sched_task_exit() */
+    sf->ra = (uint32_t)sched_task_exit;
+
+    return (char *)stk_top;
+}
+
+void IRAM_ATTR thread_yield_isr(void* arg)
+{
+    (void)arg;
+    /**
+     * Context switches are realized using software interrupts since interrupt
+     * entry and exit functions are used save and restore complete
+     * context.
+     */
+    /* clear the interrupt first */
+    WRITE_PERI_REG(SYSTEM_CPU_INTR_FROM_CPU_0_REG, 0);
+    /* set the context switch flag (indicates that context has to be switched
+       is switch on exit from interrupt in rtos_int_exit */
+    sched_context_switch_request = 1;
+}
+
+void IRAM_ATTR thread_yield_higher(void)
+{
+    /* reset hardware watchdog */
+    system_wdt_feed();
```

Comment:
Is the watchdog always on?
I would imagine otherwise that it should be the application's responsibility to feed the watchdog (using the [watchdog API](https://doc.riot-os.org/group__drivers__periph__wdt.html)) since that is what's being monitored usually.

Reply:
> Is the watchdog always on?

Yes, we have a system watchdog running all the time to avoid complete system blocking due to crashes or errors in ESP-IDF libraries, in the RIOT port or even in the application.It was implemented in the RIOT port long before the watchdog peripheral driver was added. The application can't control watchdogs since the watchdog peripheral driver is not yet supported.